sdlib/d/gc/arena.d

```diff
@@ -374,6 +428,29 @@ private:
 		unusedExtents.insert(e);
 	}
 
+	void shrinkAllocImpl(Extent* e, uint n, uint oldPages, uint delta) {
+		assert(mutex.isHeld(), "Mutex not held!");
+		assert(oldPages > 0 && oldPages <= PageCount,
+		       "Invalid number of old pages!");
+		assert(delta != 0, "Invalid delta!");
+		assert(n <= PageCount - oldPages, "Invalid index!");
+
+		auto hpd = e.hpd;
+		auto prevIndex = getFreeSpaceClass(hpd.longestFreeRange);
+
+		hpd.shrinkAlloc(n + oldPages - delta, delta);
+
+		auto index = getFreeSpaceClass(hpd.longestFreeRange);
+		if (index != prevIndex) {
+			heaps[prevIndex].remove(hpd);
+			filter &= ~(ulong(heaps[prevIndex].empty) << prevIndex);
+
+			if (!hpd.full) {
```

Comment:
hpd is necessarily not full here, because page have been freed in it. However, it might be worth checking if it is empty in an assert.